### ISSUE TITLE
Export View as Image - Scale > 100% cuts off text

### ISSUE THREAD

**Issue by smileham** [archimatetool/archi#621]
### Version of Archi
4.6 (and 4.7 beta 1)

### Operating System
Windows 10 (also seen in Windows 7)

### Expected Behaviour
Export View as Image to Raster Format, all ArchiMate component labels are visible.

100% export
![Key (Basic) - 100%](https://user-images.githubusercontent.com/2892935/81649182-d405b280-9427-11ea-81bd-f2e63baa66fc.png)

### Actual Behaviour
Export View as Image to Raster Format, however, all text elements are trimmed on the right hand side

200% export
![Key (Basic) - 200%](https://user-images.githubusercontent.com/2892935/81649180-d36d1c00-9427-11ea-876a-11aeb713d7ae.png)

### Steps to Reproduce the Behaviour
1. Export View as Image
2. Format PNG/BMP/JPEG
3. Set scale to 400%
4. View is exported to Raster Format, however, all text elements are trimmed on the right hand side

Note: Same issue appears when scaling the interface in the view.

**Comment by Phillipus** [comment]
Are you using a particular font for the elements where the text is trimmed?

Are you using a hi-res monitor?

**Comment by smileham** [comment]
Seems to happen on any font > size 9, this font is Roboto, but the same happens with many other fonts.

Not a High Res monitor

**Comment by jbsarrodie** [comment]
This is an issue on the underlaying drawing engine (draw2d) when used with a zoom other than 100%.

Zoom has always had this impact on fonts been zoomed a bit more and then cropped on the right. 

When exporting at zoom factor other than 100% I've always had such glitches (that's why I always work at 100%).

@Phillipus This is really annoying but I tried several time to fix it without success :-(

**Comment by Phillipus** [comment]
Yes, I see the same thing with some fonts (but not all).

If you increase the zoom level in the Archi View to 200% or more you'll also see the same thing.

This is a problem with some fonts on some platforms. You might see clipping on Arial font @ 200% on Windows but not on Mac, for example. Some fonts are OK, while some are not, depending on zoom level and OS.

Unfortunately, the rendering is done by the Eclipse Draw2D library (which is why you will sometimes see clipping at a greater zoom level in Archi)

**Comment by Phillipus** [comment]
> This is an issue on the underlaying drawing engine (draw2d) when used with a zoom other than 100%.

You posted at the same time as me.

>  but I tried several time to fix it without success

I would put my money on `org.eclipse.draw2d.ScaledGraphics`.

**Comment by jbsarrodie** [comment]
> You posted at the same time as me.

I was first ;-)

> This is a problem with some fonts on some platforms. You might see clipping on Arial font @ 200% on Windows but not on Mac, for example. Some fonts are OK, while some are not, depending on zoom level and OS.

Interesting, I never tested this on different OS, this might help to understand the issue...



> I would put my money on `org.eclipse.draw2d.ScaledGraphics`.

Yes, That's where I looked at...

**Comment by jbsarrodie** [comment]
FWIW, thee is a similar issue when rendering in SVG in some occasions (text clipping area too short).

**Comment by Phillipus** [comment]
> I was first ;-)

Only because I took a sip of my tea before posting.

> I would put my money on `org.eclipse.draw2d.ScaledGraphics`

There's lots of font handling going on in that class. Even if one were to extend and fix that class to use on image export it is used internally by Draw2D for the editor drawing.

**Comment by jbsarrodie** [comment]
> There's lots of font handling going on in that class. Even if one were to extend and fix that class to use on image export it is used internally by Draw2D for the editor drawing.

But if we understand the issue, maybe we would be able to change the font size we use depending on the zoom factor to compensate ;-)

**Comment by Phillipus** [comment]
> But if we understand the issue

We need a maths/graphics guru... ;-)

**Comment by jbsarrodie** [comment]
![image](https://user-images.githubusercontent.com/5757396/81660389-fbb24600-943a-11ea-9c02-e60059f15437.png)

**Comment by jbsarrodie** [comment]
> We need a maths/graphics guru... ;-)

You need Jaiguru ;-)

**Comment by jbsarrodie** [comment]
@Phillipus do you know if simple GEF examples (like the "Logic Diagram") exibit the same issue ?

**Comment by Phillipus** [comment]
> do you know if simple GEF examples (like the "Logic Diagram") exibit the same issue ?

It depends whether it uses a `ScalableFreeformLayeredPane` or a `ScalableLayeredPane`. These use scaled graphics, not sure if Logic Diagram does (I can't even get it to run as there is a bug in GEF examples plugin)

**Comment by jbsarrodie** [comment]
I've done several tests, and it seems that when the label runs on several lines, each line gets streched horizontaly with a factor which seems random:

This is a superposition of the same figure at 100% and 200%:
![image](https://user-images.githubusercontent.com/5757396/81689142-54450b80-945a-11ea-87c7-9af14fa2c403.png)

If I shrink the zoomed figure so that the first line superposes here is what I get:
![image](https://user-images.githubusercontent.com/5757396/81689216-6030cd80-945a-11ea-8a28-341ef2ffe3b4.png)

And if I try to make the bottom line to match:
![image](https://user-images.githubusercontent.com/5757396/81689353-7dfe3280-945a-11ea-8d5d-fffa74837d0e.png)

Maybe an issue with individual labels in the context of a TextFlow/BlockFlow...

**Comment by jbsarrodie** [comment]
So to summarize, font seems to get streched horizontally only (not vertically).

Could it be a side effect of `FontFactory.getAdjustedWindowsFont()`

(all my tests are done on Windows)

**Comment by Phillipus** [comment]
> Could it be a side effect of FontFactory.getAdjustedWindowsFont() ?

That only adjusts the font if the user has set Windows to non-100% scale.

**Comment by Phillipus** [comment]
> Maybe an issue with individual labels in the context of a TextFlow/BlockFlow...

BTW - latest commits in text-render branch eliminate the need for a BlockFlow figure and has simplified Text figures.

Perhaps the way to proceed is to create a class that extends `org.eclipse.draw2d.ScaledGraphics` and use it in `DiagramUtils#createModelReferencedImage()` :

```
if(scale != 1) {
   graphics = new ScaledGraphics(swtGraphics); // Use new class here
   graphics.scale(scale);
}
```

Then see what's happening.

**Comment by Phillipus** [comment]
@jbsarrodie Actually it might have something to do with TextFlow figure. A Sketch Actor figure text in a Sketch is not clipping. That uses a `org.eclipse.draw2d.Label`.

Also worth checking relationship labels before and after recent changes (used to be a Label now is a TextFlow)

**Comment by Phillipus** [comment]
> Also worth checking relationship labels before and after recent changes (used to be a Label now is a TextFlow)

Still the same.

Test case is:
- Serving relationship
- Roboto 10 regular font
- "Hello World" as text
- Zoom in diagram > 100%

![Image 1](https://user-images.githubusercontent.com/600504/81693206-7ee19500-9457-11ea-9fcd-a3b5576b03d0.png)

**Comment by Phillipus** [comment]
1. A bug in `org.eclipse.draw2d.ScaledGraphics`?
1. Something to do with TextFlow/FlowPage/Paragraphing?
1. But text is also clipping when using a `org.eclipse.draw2d.Label` on a connection

**Comment by Phillipus** [comment]
@jbsarrodie Edit `com.archimatetool.editor.diagram.figures.elements.BusinessActorFigure` constructor to use `LABEL_CONTROL` instead of `TEXT_FLOW_CONTROL` and there is no clipping.

**Comment by jbsarrodie** [comment]
> @jbsarrodie Edit `com.archimatetool.editor.diagram.figures.elements.BusinessActorFigure` constructor to use `LABEL_CONTROL` instead of `TEXT_FLOW_CONTROL` and there is no clipping.

Yes, but not for a good reason: text is still streched but not clipped by the TextFlow (which is created with its own margin and thus clipping the internal text elements)

**Comment by jbsarrodie** [comment]
> Yes, but not for a good reason: text is still streched but not clipped by the TextFlow (which is created with its own margin and thus clipping the internal text elements)

That being said, this might lead to a kind of workaround: changing the way the margin are set so that clipping only occures on the Figure. This won't fix the root cause, but few people will notice the issue ;-)

**Comment by Phillipus** [comment]
> That being said, this might lead to a kind of workaround: changing the way the margin are set so that clipping only occures on the Figure. This won't fix the root cause, but few people will notice the issue ;-)

Not sure what what you mean? (BTW best to work on figures in `text-render` branch because of recent changes)

**Comment by Phillipus** [comment]
Tried this on Mac (10.15.4) and could not get the text to clip at zoom levels > 100% with several fonts and font sizes. Have not tried with Linux.

**Comment by jbsarrodie** [comment]
> Have not tried with Linux.

I did and have the same issue, so this is not OS specific.

**Comment by jbsarrodie** [comment]
We're not the only one with this bug: https://www.eclipse.org/forums/index.php/t/154575/

(reading it and related issues at the moment)

**Comment by Phillipus** [comment]
> We're not the only one with this bug: https://www.eclipse.org/forums/index.php/t/154575/

That suggests the problem is in `TextFlow`, however I can see the same problem when we were using a `Label` in `AbstractDiagramConnectionFigure` (above https://github.com/archimatetool/archi/issues/621#issuecomment-627322111)

**Comment by Phillipus** [comment]
Experimented with the version of `com.archimatetool.editor.diagram.figures.connections.AbstractDiagramConnectionFigure` in `master` branch that uses a `Label`

Over-ride `Label` to add an extra 10 pixels to the text width:

```
fConnectionLabel = new Label("") {
    protected Dimension calculateTextSize() {
        Dimension d = super.calculateTextSize();
        d.width += 10;
        return d;
    }
};
```

Fixes the issue in https://github.com/archimatetool/archi/issues/621#issuecomment-627322111

Note that `Label#calculateTextSize()` is this:

```
protected Dimension calculateTextSize() {
	return getTextUtilities().getTextExtents(getText(), getFont());
}
```

So perhaps the issue is in `org.eclipse.draw2d.TextUtilities`

**Comment by Phillipus** [comment]
I would say that the issue is in `org.eclipse.draw2d.TextUtilities`.

However, for a `TextFlow` it is indirectly accessed via `org.eclipse.draw2d.text.FlowUtilities`

So, the following hack works:

```
TextFlow textFlow = new TextFlow() {
    @Override
    protected FlowUtilities getFlowUtilities() {
        FlowUtilities fu = new FlowUtilities() {
            @Override
            protected TextUtilities getTextUtilities() {
                TextUtilities tu = new TextUtilities() {
                    @Override
                    public Dimension getTextExtents(String s, Font f) {
                        Dimension d = super.getTextExtents(s, f);
                        d.width += 8;
                        return d;
                    }
                };
                return tu;
            }
        };
        return fu;
    }
};
```

This adds on an extra 8 pixels to the text extends width in `getTextExtents()`. So maybe this should take into account the zoom factor?

**Comment by Phillipus** [comment]
Actually we should look at `org.eclipse.draw2d.FigureUtilities` because in `org.eclipse.draw2d.TextUtilities`:

```
public Dimension getTextExtents(String s, Font f) {
	return FigureUtilities.getTextExtents(s, f);
}
```

Which is this:

```
protected static org.eclipse.swt.graphics.Point getTextDimension(String s,
		Font f) {
	setFont(f);
	return getGC().textExtent(s);
}
```

**Comment by Phillipus** [comment]
My money is on `org.eclipse.draw2d.FigureUtilities` and zoom factor.

**Comment by Phillipus** [comment]
Here's a hack:

```
    static FlowUtilities flowUtilities = new FlowUtilities() {
        @Override
        protected TextUtilities getTextUtilities() {
            return textUtilities;
        }
    };
    
    static TextUtilities textUtilities = new TextUtilities() {
        @Override
        public Dimension getTextExtents(String s, Font f) {
            Dimension d = super.getTextExtents(s, f);
            d.width *= 1.1;
            return d;
        }
    };
    
    TextFlow textFlow = new TextFlow() {
        @Override
         protected FlowUtilities getFlowUtilities() {
            return flowUtilities;
         }
    };



```

**Comment by Phillipus** [comment]
@jbsarrodie There's a new branch with the hack in it `text-hack`

**Comment by jbsarrodie** [comment]
You've ended up with almost the same hack as me yesterday evening (mine also shrink a bit the `Graphics` before drawing the label at zoom factors other than 100%).

But these are hacks and and don't always work (I have seen cases were the width difference is more than 10%).

In fact the issue lies in the way `org.eclipse.draw2d.FigureUtilities#getTextExtents()` is called and implemented:
- It seems to be called with the font height at 100%, assuming that drawing a string at other zoom factors will take `size_at_100% * zoom factor` space, while in fact it isn't always the case because `ScaledGraphics` does weird computations on font height and casts it to integer.
- In addition, for its size calculation, `org.eclipse.draw2d.FigureUtilities` uses a private `GC` which is not set exactly the same way as the GC used by the Editor (no advanced graphics and maybe no anti-aliasing too), which can also lead to some differences in font rendering, thus text size.

So the path I'm following (and I have to experiment) is to find a better way to implement `org.eclipse.draw2d.FigureUtilities#getTextExtents()` so that it takes the zoom factor into account:
- instead of computing the size with a font height `h` (so independent of zoom factor)...
- compute the size of a string with a font height `scaled_h` (being the exact same font height calculated by `ScaledGraphics`) and divide it by current zoom factor.

**Comment by Phillipus** [comment]
How are you getting the zoom factor? `graphics#getAbsoluteScale()` ?

**Comment by jbsarrodie** [comment]
> How are you getting the zoom factor? `graphics#getAbsoluteScale()` ?

Yes, that's how I get it (don't know if there's better option or side effects).

**Comment by Phillipus** [comment]
That's the only way to get it, really, at the point of drawing the graphics.

**Comment by jbsarrodie** [comment]
Re font zooming in `ScaledGraphics` this seems to be as simple as `return (int) (zoom * height)` (see `zoomFontHeight`), but there might be some hiddent computation on GC.

FWIW:
- `GC#stringExtent()` does some DPI related computation, maybe this also has some side effects.
- `ScaledGraphics#drawText()`, `ScaledGraphics#fillText()` and `ScaledGraphics#fillString()` rely on `ScaledGraphics#zoomTextPoint()` which does weird computations based on font metrics and `fractionalX` & `fractionalY` which are not very clear.

BTW: What I still really don't know is if it is "normal" for fonts to be a bit stretched or shrinked at some font size (in such case, texte layout **is** impacted by zoom factor and in some occasion TextFlow will have to change the layout based on it), or if this is only an eclipse (or swt) issue (in such case it could be needed to dynamically change the font height depending on the zoom factor so that text layout is not impacted).

**Comment by Phillipus** [comment]
> 
> 
> ![image](https://user-images.githubusercontent.com/5757396/81660389-fbb24600-943a-11ea-9c02-e60059f15437.png)

**Comment by jbsarrodie** [comment]
> So the path I'm following (and I have to experiment) is to find a better way to implement `org.eclipse.draw2d.FigureUtilities#getTextExtents()` so that it takes the zoom factor into account

I have an issue for this: how can I get the current `ScaledGraphics` object? Any idea?

**Comment by Phillipus** [comment]
> I have an issue for this: how can I get the current ScaledGraphics object? Any idea?

Where are you trying to get it? `if(graphics instanceof ScaledGraphics)` ?

**Comment by jbsarrodie** [comment]
> Where are you trying to get it? `if(graphics instanceof ScaledGraphics)` ?

I'm in our patched `TextUtilities#getTextExtents()` called by our patched `TextFlow#getFlowUtilities()` called by...

So I'm deep in the code called each time the layout has to be recomputed before painting the figure (`Figure#layout()`).

Looking at the stack trace, maybe I can get the zoom factor from `ScalableFreeformLayeredPane`:
![image](https://user-images.githubusercontent.com/5757396/81930294-ad42aa00-95e8-11ea-9bbe-9341d0c12ca7.png)

So my only idea atm is to use `Figure.getParent()` until I reach one which implements `org.eclipse.draw2d.ScalableFigure`...

**Comment by Phillipus** [comment]
But you can't use the zoom factor as set in the editor, because it won't work when exporting the image at different scales (that doesn't take the zoom factor into account)

**Comment by Phillipus** [comment]
Look at `DiagramUtils#createModelReferencedImage()` it is independent of the GEF Zoom Manager.

**Comment by jbsarrodie** [comment]
> But you can't use the zoom factor as set in the editor, because it won't work when exporting the image at different scales (that doesn't take the zoom factor into account)

Yes, but my first step it to test and see if my idea works. If it does, then I'll see how to do it the right way.

**Comment by jbsarrodie** [comment]
Still investigating...

Potentially useful references:
- https://bugs.eclipse.org/bugs/show_bug.cgi?id=470334
- https://bugs.eclipse.org/bugs/show_bug.cgi?id=442442

In short, `ScaledGraphics` is deprecated, but someone shared a potential workaround.

**Comment by Phillipus** [comment]
"and replace its usages with direct scaling of the underlying SWTGraphics."

Interesting.

(Go and drink some wine or something ;-) )

**Comment by jbsarrodie** [comment]
> Interesting

Just tested but doesn't work. Still investigating...

**Comment by jbsarrodie** [comment]
> Go and drink some wine or something ;-)

I can't (OCD)

**Comment by jbsarrodie** [comment]
> Just tested but doesn't work. Still investigating...

I first thought that I could apply the [suggested workaround](https://bugs.eclipse.org/bugs/show_bug.cgi?id=470334) in `AbstractDiagramEditor#createGraphicalViewer()` because that's already there you started to patch `ScrollingGraphicalViewer` which is the "entry point" of the workaround (here I'm using updated code from `ScalableLayeredPane`):

````java
        final GraphicalViewer viewer = new ScrollingGraphicalViewer() {
        	// Start of patch
        	protected void createDefaultRoot() {
        		setRootEditPart(new ScalableRootEditPart() {
    			protected ScalableLayeredPane createScaledLayers() {
    				ScalableLayeredPane layers = new ScalableLayeredPane() {
    					protected void paintClientArea(Graphics graphics) {
    						if (getChildren().isEmpty())
    							return;
    						if (getScale() == 1.0) {
    							super.paintClientArea(graphics);
    						} else {
//        							ScaledGraphics g = new ScaledGraphics(graphics);
    							Graphics g = graphics;
    							boolean optimizeClip = getBorder() == null
    									|| getBorder().isOpaque();
    							if (!optimizeClip)
    								g.clipRect(getBounds().getCropped(getInsets()));
    							g.scale(getScale());
    							g.pushState();
    							paintChildren(g);
    							g.dispose();
    							graphics.restoreState();
    						}
    					}

    				};
        				layers.add(createGridLayer(), GRID_LAYER);
        				layers.add(getPrintableLayers(), PRINTABLE_LAYERS);
//        				layers.add(new FeedbackLayer(), SCALED_FEEDBACK_LAYER);
        				return layers;
        			}
        		});
        	}
        	// End of patch
            @Override
            protected LightweightSystem createLightweightSystem() {
            [...]
````

But this didn't work. In fact this didn't work because this patched `ScalableLayeredPane#paintClientArea()` is never called. I guess the real one is set somewhere else in our codebase but I still haven't found it (some help would be apreciated).

But then I remembered one of your comment: we can easily try any workaround in `DiagramUtils#createModelReferencedImage()`, so that's what I did. I simply commented the line:
````java
graphics = new ScaledGraphics(swtGraphics);
````
So that we end up with:
````java
        // If scaled, then scale now
        if(scale != 1) {
//            graphics = new ScaledGraphics(swtGraphics);
            graphics.scale(scale); // SWTGraphics supports scale()
        }
````

And this works!

My reference image (view exported at 300%) in Archi 4.7 beta1:
![Applications_300](https://user-images.githubusercontent.com/5757396/82022508-154acc00-968d-11ea-813c-7c0b8a2fbdfd.png)

And the same after commenting `ScaledGraphics`:
![Applications_300_fix](https://user-images.githubusercontent.com/5757396/82022556-27c50580-968d-11ea-83b5-f6ea5516b449.png)

I now have to find where to do the same for the whole Editor.

**Comment by Phillipus** [comment]
> I simply commented the line:
> `graphics = new ScaledGraphics(swtGraphics);`

Wow. I don't even know why `ScaledGraphics` was used. Must have been taken from an example code.

**Comment by Phillipus** [comment]
Re: `ScalableLayeredPane`

this is used in `ScalableRootEditPart` and that is used in `GraphicalViewerImpl#createDefaultRoot()`

**Comment by Phillipus** [comment]
> But this didn't work. In fact this didn't work because this patched `ScalableLayeredPane#paintClientArea()` is never called.

If you put a break point in the original unpatched `ScalableLayeredPane#paintClientArea()` code you'll see it is not called. In fact, the only one of its methods that seems to get called is `translateToParent()`

**Comment by Phillipus** [comment]
Perhaps the work is being done in one of the scaled layers in `ScalableRootEditPart`?

```

	protected LayeredPane createPrintableLayers() {
		LayeredPane pane = new LayeredPane();

		Layer layer = new Layer();
		layer.setLayoutManager(new StackLayout());
		pane.add(layer, PRIMARY_LAYER);

		layer = new ConnectionLayer();
		layer.setPreferredSize(new Dimension(5, 5));
		pane.add(layer, CONNECTION_LAYER);

		return pane;
	}

```

**Comment by Phillipus** [comment]
`ScalableFreeformLayeredPane`?

**Comment by Phillipus** [comment]
Possibly want to hook in at `AbstractDiagramEditor#createRootEditPart()' which is implemented in all three sub-classes.

**Comment by Phillipus** [comment]
`ScalableFreeformLayeredPane#paintClientArea` uses `ScaledGraphics`

**Comment by jbsarrodie** [comment]
> `ScalableFreeformLayeredPane#paintClientArea` uses `ScaledGraphics`

But is never called..

**Comment by Phillipus** [comment]
> But is never called..

It is when you zoom in. I set a breakpoint and it is called.

**Comment by jbsarrodie** [comment]
Sorry, I misread the classname. Looking at it...

**Comment by Phillipus** [comment]
I've done it. Fixed it. :-)

**Comment by Phillipus** [comment]
Incomng fix...

**Comment by jbsarrodie** [comment]
> I've done it. Fixed it. :-)

Created from `ArchimateDiagramEditor#createRootEditPart()`...

**Comment by Phillipus** [comment]
yep. I just need to check whether to `graphics.pushState();` or not...

**Comment by Phillipus** [comment]
I need to create a re-usable subclass and put it in `AbstractDiagramEditor#createRootEditPart()`

I'll commit something soon...

**Comment by jbsarrodie** [comment]
BTW, in the meantime, I found what should be the simpliest way to set a borderwidth to figures without issues...

**Comment by Phillipus** [comment]
> BTW, in the meantime, I found what should be the simpliest way to set a borderwidth to figures without issues...

Perhaps we should open another issue in case this one turns into a mega-thread... ;-)

**Comment by jbsarrodie** [comment]
> Perhaps we should open another issue in case this one turns into a mega-thread... ;-)

Of course yes (that's the reason why I didn't add more info). But I have to see if we already have an issue for that (I'm sure we have). Will look at that later today.

**Comment by Phillipus** [comment]
I've committed it now.

`com.archimatetool.editor.diagram.editparts.ExtendedScalableFreeFormRootEditPart`

I'm not 100% certain about whether we need `graphics.pushState();` but I think it all works.

**Comment by Phillipus** [comment]
Actually I think we also need a popState().

**Comment by Phillipus** [comment]
Sketch View is doing something weird when zooming...

**Comment by Phillipus** [comment]
I've committed a tweak (unrelated to Sketch View problem)

But if you create a Sketch View only the background image is zoomed. This could indicate a deeper problem.

**Comment by Phillipus** [comment]
@jbsarrodie There's problem. If the grid is displayed, zoom doesn't work.

**Comment by Phillipus** [comment]
@jbsarrodie Fixed it. As I thought:

> I'm not 100% certain about whether we need graphics.pushState();

**Comment by Phillipus** [comment]
@jbsarrodie We need to be certain about this push/pop/restore state. Because in the original code there was two Graphics objects and now we have only one.

**Comment by Phillipus** [comment]
Wow, that was some detective work. I think (hope) we nailed it now.

**Comment by Phillipus** [comment]
It might have been quicker if I had followed the link (https://bugs.eclipse.org/bugs/show_bug.cgi?id=470334) in the comment above and looked at the example code. ;-) (But the code is much the same)

**Comment by Phillipus** [comment]
If only they had fixed this issue - https://bugs.eclipse.org/bugs/show_bug.cgi?id=442442

;-)

**Comment by jbsarrodie** [comment]
> It might have been quicker if I had followed the link (https://bugs.eclipse.org/bugs/show_bug.cgi?id=470334) in the comment above and looked at the example code. ;-) (But the code is much the same)

You didn't ? !

That was the key thing yesterday evening for me. When I saw this I knew we would solve it.

> If only they had fixed this issue - https://bugs.eclipse.org/bugs/show_bug.cgi?id=442442

I wonder if it wouldn't be more simple for us to fork draw2d and GEF to solve all those issues (not only this one, but also the other we had in the past) at their core.

**Comment by Phillipus** [comment]
> You didn't ? !

I skim read your message and narrowed it down to `ScalableFreeformLayeredPane` and then saw that it was using `ScaledGraphics` and hacked at it. If I had slowed down and read your links and posts properly I would have done it quicker and avoided the popState() issue.

> I wonder if it wouldn't be more simple for us to fork draw2d and GEF to solve all those issues (not only this one, but also the other we had in the past) at their core.

Yep, I was thinking the same. I don't feel like doing that right now (I think it might involve some issues) but I shall certainly investigate it.

**Comment by jbsarrodie** [comment]
> Yep, I was thinking the same. I don't feel like doing that right now (I think it might involve some issues) but I shall certainly investigate it.

I'm sure that, if we do it, we'll have lot's of people relying on our work ;-) Maybe I could talk to our Obeo friends to see if there is a more "official" way of updating the codebase....

**Comment by Phillipus** [comment]
What I find extraordinary is the lack of maintenance of GEF/Drawd2D considering how many apps use it. And not only that, how many stacks build on it - GMF, Sirius...

**Comment by jbsarrodie** [comment]
> Wow, that was some detective work, JB.

![image](https://user-images.githubusercontent.com/5757396/82069447-9927a700-96d3-11ea-9b1d-d5665cf5abe0.png)

**Comment by Phillipus** [comment]
@jbsarrodie 

Actually, rather than hacking the GEF/Draw2D code why don't we just remove all the Zoom and image scaling features from Archi? It would be easier...
<p>&nbsp;</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
<p>&nbsp;</p>
...joking! ;-)

**Comment by Phillipus** [comment]
Committed another fix for this. Connection dashes/dots spacing don't need scaling with the zoom hack above 1.0 scale.

**Comment by jbsarrodie** [comment]
> Committed another fix for this. Connection dashes/dots spacing don't need scaling with the zoom hack above 1.0 scale.

This makes me think that there are some zoom related code in orthogonal anchor. I have to have a look at it and do some tests.

**Comment by jbsarrodie** [comment]
> This makes me think that there are some zoom related code in orthogonal anchor. I have to have a look at it and do some tests.

Everything seems ok :-)

**Comment by Phillipus** [comment]
That `ScaledGraphics` class really is a piece of sh*t, isn't it?

**Comment by jbsarrodie** [comment]
> That `ScaledGraphics` class really is a piece of sh*t, isn't it?

Yes, and I could rewrite it from scratch quickly....

````java
public class ScaledGraphics extends SWTGraphics {
	public ScaledGraphics (GC gc) {
		super(gc);
	}
}
````

**Comment by Phillipus** [comment]
Yes!!!!

@jbsarrodie Do you remember the thumbnail bug on Mac (Mojave and later versions)? Last year I bought a MacBook specifically to solve the problem. 

The issue is that, for some reason, on Mac the `org.eclipse.swt.graphics.GC` instance does not flush the previous tiled image when building up the thumbnail so the thumbnail has repeated tiled images.

Last year I wrote a hack to copy the tiled image and re-draw it onto the `GC`, but this hack stopped working after Eclipse 4.13.

So I wrote a new hack that simply used bigger tiled images on Mac. It kind of worked, but not always.

Then, today, I remembered that our hacked Thumbnail class (`org.eclipse.draw2d.archi.parts.Thumbnail`) uses `ScaledGraphics` so I removed that instance. This didn't fix the thumbnail issue but it did mean that I looked closer at the `GC` and `SWTGraphics` instances.

So, the hack is, for Mac - each time the tiled image is drawn dispose of the previous `GC` and `SWTGraphics` instances and create new ones. Then the tiled image is re-drawn.

Yesss!

TL;DR - this is nerdy stuff.

**Comment by Phillipus** [comment]
@jbsarrodie I just tested the thumbnail on Linux and it's not working in either the latest Archi code, nor 4.5.1, nor 4.6. I'm using Linux Mint (Ubuntu) with screen resolution set at hi-res. Can you check whether it works on your Linux?

**Comment by jbsarrodie** [comment]
> @jbsarrodie I just tested the thumbnail on Linux and it's not working in either the latest Archi code, nor 4.5.1, nor 4.6. I'm using Linux Mint (Ubuntu) with screen resolution set at hi-res. Can you check whether it works on your Linux?

You mean this:
![Screenshot from 2020-05-16 13-35-58](https://user-images.githubusercontent.com/5757396/82118877-43f2a080-977a-11ea-8435-3a805f253ad5.png)

?

(done with Archi 4.7 EA ?)

**Comment by Phillipus** [comment]
@jbsarrodie Thanks for checking. Also works for me. Except for hi-res. 

See https://github.com/archimatetool/archi/issues/624

**Comment by Phillipus** [comment]
@smileham This should be fixed in Archi 4.7 beta 3

**Comment by smileham** [comment]
Wow, that seems quite the rabbit hole you guys went down!  I felt guilty at first, but following the thread, it looks like you both seem to have had fun ;)

Much appreciated, just tested, looks good!
